cms-contentful/src/main/java/io/commercetools/sunrise/cms/contentful/ContentfulCmsPage.java

```diff
@@ -0,0 +1,137 @@
+    }
+
+    private CDAEntry getEntryWithContentField(@Nullable final CDAEntry cdaEntry, final FieldIdentifier fieldIdentifier) {
+        if (fieldIdentifier.isLastLevelEntry()) {
+            return cdaEntry;
+        } else {
+            final String key = fieldIdentifier.getCurrentEntryName();
+            if (cdaEntry != null && cdaEntry.rawFields().containsKey(key)) {
+                Object item = cdaEntry.getField(key);
+                if (item instanceof CDAEntry) {
+                    fieldIdentifier.removeCurrentEntryName();
+                    return getEntryWithContentField((CDAEntry) item, fieldIdentifier);
+                }
+            }
+            return null;
+        }
+    }
+
+    // CDAField contains information related to field, like it's type
+    private Optional<CDAField> getCdaField(final CDAEntry lastLevelEntry, final String fieldName) {
+        return lastLevelEntry.contentType().fields()
+                .stream()
+                .filter(cdaField -> cdaField.id().equals(fieldName)
+                        && FieldTypes.ALL_SUPPORTED.contains(cdaField.type()))
+                .findFirst();
+    }
+    
+    private Object getLocalizedEntryField(final Locale locale, final CDAEntry lastLevelEntry, final String fieldName) {
+        lastLevelEntry.setLocale(locale.toLanguageTag());
+        return lastLevelEntry.getField(fieldName);
+    }
+
+    private Optional<Locale> getFirstSupportedLocale(final CDAEntry lastLevelEntry, final String fieldName) {
+        final Map<String, Object> rawFields = lastLevelEntry.rawFields();
+        final Map<String, Object> localeContentMap = getLocaleContentMap(fieldName, rawFields);
+        return Optional.ofNullable(localeContentMap).flatMap(map -> {
+            final Set<String> allSupportedLocales = map.keySet();
+            return locales.stream()
+                    .filter(locale -> allSupportedLocales.contains(locale.toLanguageTag()))
+                    .findFirst();
+        });
+    }
+
+    private Optional<String> getContentAccordingToFieldDefinition(@Nullable final Object localizedEntryField,
+                                                                  final CDAField cdaField) {
+        return Optional.ofNullable(localizedEntryField)
+                .map(entryField -> {
+                    String content = null;
+                    if (FieldTypes.CONVERTABLE_TO_STRING.contains(cdaField.type())) {
+                        content = String.valueOf(entryField);
+                    } else if (cdaField.linkType().equals(FieldTypes.LINK_ASSET)) {
+                        content = ((CDAAsset) entryField).url();
+                    }
+                    return content;
+                });
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> getLocaleContentMap(String fieldName, Map<String, Object> rawFields) {
+        return (Map<String, Object>) rawFields.get(fieldName);
+    }
+
+    private class FieldIdentifier {
```

Comment:
It's a wild guess because I haven't checked, but I think you can use an `[Iterator](https://docs.oracle.com/javase/7/docs/api/java/util/Iterator.html)` to do most of this methods (`hasNext`, `next`, `remove`). Why did you need to implement your own class?